[SRVKS-1115] Performance and scalability of OpenShift Serverless Serving (#80197)
openshift/openshift-docs: modules/serverless-config-high-workloads-serving.adoc

====

To fine-tune your installation, use the above findings combined with your own test results to configure the `KnativeServing` custom resource:

[source,yaml]
----
apiVersion: operator.knative.dev/v1beta1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  high-availability:
    replicas: 2 <1>
  workloads:
    - name: component-name <2>
      replicas: 2 <2>
      resources:
        - container: container-name
          requests:
            cpu: <3>
            memory: <3>
          limits:
            cpu: <3>
            memory: <3>
  podDisruptionBudgets: <4>
    - name: name-of-pod-disruption-budget
      minAvailable: 1
----
<1> Set this to at least 2, to make sure you have always at least two instances of every component running. You can also use `workloads` to override the replicas for certain components.

> <2> Use the `workloads` list to configure specific components. Use the `deployment` name of the component (like `activator`, `autoscaler`, `autoscaler-hpa`, `controller`, `webhook`, `net-kourier-controller`, `3scale-kourier-gateway`, `net-istio-controller`) and set the `replicas`.

Reviewer: We should also add a note that would say something like:

For components that use HorizontalPodAutoscalers  (activator, webhook and 3scale-kourier-gateway) , the `replicas` field sets the _minimum_ number of replicas. Actual number of replicas will depend on the CPU load and scaling done by the HPAs.

<3> Set the requested and limited CPU + memory according to at least the idle consumption while also taking the above findings and your own test results into consideration.

> <4> Adjust the `PodDistruptionBudgets` to a value lower or equal to the `replicas`, to avoid problems during node maintenance. The default `minAvailable` is set to `1`, so if you increase the desired replicas, make sure to also increase `minAvailable`.

Reviewer: I think we should rather recommend just "lower to the replicas", not equal.  Setting it to equal would generate a PodDisruptionBudgetAtLimit alert, 

https://github.com/openshift/runbooks/blob/master/alerts/cluster-kube-controller-manager-operator/PodDisruptionBudgetAtLimit.md

> Standard workloads should have at least one pod more than is desired to support [API-initiated eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/api-eviction/). Workloads that are at the minimum disruption allowed level violate this and could block node drain. This is important for node maintenance and cluster upgrades.

[IMPORTANT]
====
Fast startup and fast response-times of your Knative Service user workloads are critical for good performance of the overall system. The Knative Serving components are buffering incoming requests when the Knative Service user backend is scaling-up or when request concurrency has reached its capacity. If your Knative Service user workload introduces long startup or request-latency, it will either overload the `activator` component (when the CPU and memory configuration is too low) or lead to errors for the calling clients.
====

.Procedure

* To fine-tune your installation, use the previous findings combined with your own test results to configure the `KnativeServing` custom resource:
+
.A high workload configuration in KnativeServing CR
[source,yaml]
----
apiVersion: operator.knative.dev/v1beta1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  high-availability:
    replicas: 2 <1>
  workloads:
    - name: component-name <2>
      replicas: 2 <2>
      resources:
        - container: container-name
          requests:
            cpu: <3>
            memory: <3>
          limits:
            cpu: <3>

>             memory: <3>

Reviewer 2: Duplicate callouts are not supported on docs.redhat.com, each must be a unique number. In this case, the key at the bottom is also describing multiple parameters in a single callout. It should be one-to-one.

  podDisruptionBudgets: <4>
    - name: name-of-pod-disruption-budget
      minAvailable: 1
----
<1> Set this parameter to at least `2` to make sure you always have at least two instances of every component running. You can also use `workloads` to override the replicas for certain components.

> <2> Use the `workloads` list to configure specific components. Use the `deployment` name of the component and set the `replicas` field. For components that use the HorizontalPodAutoscalers (`activator`, `webhook`, and `3scale-kourier-gateway`), the `replicas` field sets the minimum number of replicas. The actual number of replicas will depend on the CPU load and scaling done by the HPAs.  

Reviewer 3: ```suggestion
<2> Use the `workloads` list to configure specific components. Use the `deployment` name of the component and set the `replicas` field. For components that use the HorizontalPodAutoscalers (HPA) (`activator`, `webhook`, and `3scale-kourier-gateway`), the `replicas` field sets the minimum number of replicas. The actual number of replicas will depend on the CPU load and scaling done by the HPAs.  
```

<3> Set the requested and limited CPU + memory according to at least the idle consumption while also taking the above findings and your own test results into consideration.

> <4> Adjust the `PodDistruptionBudgets` to a value lower than `replicas`, to avoid problems during node maintenance. The default `minAvailable` is set to `1`, so if you increase the desired replicas, make sure to also increase `minAvailable`.

Reviewer 3: ```suggestion
<4> Adjust the `PodDistruptionBudgets` to a value lower than `replicas`, to avoid problems during node maintenance. The default `minAvailable` is set to `1`, so if you increase the required replicas, make sure to also increase `minAvailable`.
```
See ISG: https://www.ibm.com/docs/en/ibm-style?topic=word-usage#d

[IMPORTANT]
====
As each environment is highly specific, it is essential to test and find your own ideal configuration.

> Use the monitoring and alerting functionality of OpenShift to continuously monitor your actual resource consumption and make adjustments if needed.

Reviewer 2: ```suggestion
Use the monitoring and alerting functionality of {product-title} to continuously monitor your actual resource consumption and make adjustments if needed.
```

> If you are using the OpenShift Serverless and Service Mesh integration, additional CPU overhead is added by the `istio-proxy` sidecar containers.

Reviewer 3: ```suggestion
If you are using the OpenShift Serverless and Service Mesh integration, additional CPU processing is added by the `istio-proxy` sidecar containers.
```
See ISG: https://www.ibm.com/docs/en/ibm-style?topic=word-usage#o

> If you are using the OpenShift Serverless and Service Mesh integration, additional CPU processing is added by the `istio-proxy` sidecar containers.

Reviewer 2: ```suggestion
If you are using the {ServerlessProductName} and {SMProductShortName} integration, additional CPU processing is added by the `istio-proxy` sidecar containers.
```

> For more information on this, see the Service Mesh documentation.

Reviewer 3: ```suggestion
For more information about this, see the Service Mesh documentation.
```
See ISG: https://www.ibm.com/docs/en/ibm-style?topic=word-usage#i

> For more information about this, see the Service Mesh documentation.

Reviewer 2: ```suggestion
For more information about this, see the {SMProductShortName} documentation.
```
